diagrammer/python/scene.py

```diff
@@ -463,6 +464,66 @@ def gps(self) -> None:
         else:
             self._width = self._height = 0
 
+        # make overlapping references' path BEZIER_CLOCK
+        ref_by_angle = defaultdict(list)
+
+        for ref in self._references:
+            # get canonical degree, which is floored and [0, 180)
+            canonical_degree = math.degrees(ref.get_head_angle())
+
+            while not (0 <= canonical_degree < 180):
+                if canonical_degree < 0:
+                    canonical_degree += 180
+                elif canonical_degree >= 180:
+                    canonical_degree -= 180
+                else:
+                    raise FloatingPointError(f'Scene.gps: canonical_degree {canonical_degree} is just wrong')
+
+            canonical_degree = math.floor(canonical_degree)
+            assert type(canonical_degree) is int
+
+            ref_by_angle[canonical_degree].append(ref)
+
+        # loop through all angle lists
+        for same_angle_list in ref_by_angle.values():
+            # for each angle list, loop through all pairs of refs
+            for i in range(len(same_angle_list)):
```

Comment:
this is fine, I just thought I'd drop an alternate way to do it in case you decide you prefer that:

for i, ref1 in enumerate(same_angle_list):
    for j, ref2 in enumerate(same_angle_list[i+1:]):
            ...

although I think slices might be copies so your way avoids some overhead if that's the case but just a thought

Reply:
hmm yeah that's better design according to the problem I graded for ics33

Reply:
slicing doesn't copy actually so I'll change it to that